packages/app-mobile/components/NoteEditor/MarkdownToolbar.tsx

```diff
@@ -0,0 +1,577 @@
+			editorControl.toggleList(ListType.CheckList);
+		}, [editorControl]),
+	});
+
+
+	listButtons.push({
+		icon: (
+			<AntIcon name="indent-left" style={styles.text}/>
+		),
+		accessibilityLabel: _('Decrease indent level'),
+		onClick: editorControl.decreaseIndent,
+	});
+
+	listButtons.push({
+		icon: (
+			<AntIcon name="indent-right" style={styles.text}/>
+		),
+		accessibilityLabel: _('Increase indent level'),
+		onClick: editorControl.increaseIndent,
+	});
+
+
+	// Inline formatting
+	const inlineFormattingBtns: ButtonSpec[] = [];
+	inlineFormattingBtns.push({
+		icon: (
+			<FontAwesomeIcon name="bold" style={styles.text}/>
+		),
+		accessibilityLabel:
+			selState.bolded ? _('Unbold') : _('Bold text'),
+		active: selState.bolded,
+		onClick: editorControl.toggleBolded,
+	});
+
+	inlineFormattingBtns.push({
+		icon: (
+			<FontAwesomeIcon name="italic" style={styles.text}/>
+		),
+		accessibilityLabel:
+			selState.italicized ? _('Unitalicize') : _('Italicize'),
+		active: selState.italicized,
+		onClick: editorControl.toggleItalicized,
+	});
+
+	inlineFormattingBtns.push({
+		icon: '{;}',
+		accessibilityLabel:
+			selState.inCode ? _('Remove code formatting') : _('Format as code'),
+		active: selState.inCode,
+		onClick: editorControl.toggleCode,
+	});
+
+	if (props.editorSettings.katexEnabled) {
+		inlineFormattingBtns.push({
+			icon: '∑',
+			accessibilityLabel:
+				selState.inMath ? _('Remove TeX region') : _('Create TeX region'),
+			active: selState.inMath,
+			onClick: editorControl.toggleMath,
+		});
+	}
+
+	inlineFormattingBtns.push({
+		icon: (
+			<FontAwesomeIcon name="link" style={styles.text}/>
+		),
+		accessibilityLabel:
+			selState.inLink ? _('Edit link') : _('Create link'),
+		active: selState.inLink,
+		onClick: editorControl.showLinkDialog,
+	});
+
+
+	// Actions
+	const actionButtons: ButtonSpec[] = [];
+	actionButtons.push({
+		icon: (
+			<FontAwesomeIcon name="calendar-plus" style={styles.text}/>
+		),
+		accessibilityLabel: _('Insert time'),
+		onClick: useCallback(() => {
+			editorControl.insertText(time.formatDateToLocal(new Date()));
+		}, [editorControl]),
+	});
+
+	actionButtons.push({
+		icon: (
+			<MaterialIcon name="spellcheck" style={styles.text}/>
+		),
+		accessibilityLabel:
+			props.selectionState.spellChecking ? _('Check spelling') : _('Stop checking spelling'),
+		active: props.selectionState.spellChecking,
+		disabled: props.selectionState.unspellCheckableRegion,
+		onClick: useCallback(() => {
+			editorControl.setSpellcheckEnabled(!props.selectionState.spellChecking);
+		}, [editorControl]),
+	});
+
+	actionButtons.push({
+		icon: (
+			<MaterialIcon name="search" style={styles.text}/>
+		),
+		accessibilityLabel: _('Find and replace'),
+		active: props.searchState.dialogVisible,
+		onClick: useCallback(() => {
+			if (props.searchState.dialogVisible) {
+				editorControl.searchControl.hideSearch();
+			} else {
+				editorControl.searchControl.showSearch();
+			}
```

Comment:
I don't think this editorControl object is well setup because it has hidden dependencies, so you might end up using stale state over there. For example `showSearch` uses `searchState`, yet the editorControl is not recreated when `searchState` is changed.

I'm thinking, rather than passing editorControl to everything, you might want to pass only the specific handler needed by your callback. Then on the editorControl, you define at least the functions that have dependencies as callback. So `showSearch: useCallback(() => ...., [searchState])`

And here you pass `editorControl.searchControl.showSearch` as dependency.

Edit: now that I see the showSearch code again, I think you should call `setSearchState((prev) => { return { ...prev, dialogVisible: true })` and that way you don't need useCallback. But I still think that passing the full editorControl as dependency to everything is not ideal in terms of rendering. Each useCallback should only have a dependency to the specific handler they need.

Reply:
I've updated this with a `searchStateRef = RefObject<SearchState>`. As such, neither `searchControl` nor `editorControl` need to be re-created when `SearchState` changes.